src/main/java/ch/njol/skript/effects/EffOperations.java

```diff
@@ -0,0 +1,209 @@
+					.filter(info -> info.getRight().isAssignableFrom(operativeType))
+					.map(OperationInfo::getReturnType)
+					.toArray(Class[]::new);
+				if (returnTypes.length == 0) {
+					Skript.error(operator.getName() + " cannot be performed with "
+						+ Utils.a(getClassInfoCodeName(operativeType)));
+					return false;
+				}
+			} else {
+				returnTypes = Arithmetics.getOperations(operator, baseType).stream()
+					.map(OperationInfo::getReturnType)
+					.toArray(Class[]::new);
+				if (returnTypes.length == 0) {
+					Skript.error("This expression can not be " + getOperatorString() + ".");
+					return false;
+				}
+			}
+		} else {
+			// We've deduced that the return types from both 'base' and 'operative' are guaranteed not to be 'Object.class'
+			// We can check to see if an operation exists, if not, parse error
+			operation = getOperation(baseType, operativeType);
+			if (operation == null) {
+				Skript.error("This expression cannot be " + getOperatorString() + " by "
+					+ Utils.a(getClassInfoCodeName(operativeType)) + ".");
+				return false;
+			}
+		}
+		return LiteralUtils.canInitSafely(operative);
+	}
+
+	@Override
+	protected void execute(Event event) {
+		Object operativeObject = operative.getSingle(event);
+		if (operativeObject == null)
+			return;
+		// Ensure 'operativeType' is not 'Object.class' by using '#getReturnType' or the class of the object
+		Class<?> operativeType = operative.getReturnType().equals(Object.class) ? operativeObject.getClass() : operative.getReturnType();
+		Operation<Object, Object, Object> operation = this.operation;
+		Function<?, Object> changerFunction = null;
+		Class<?> baseType = base.getReturnType();
+		// Convert array classes to singular, allowing for easier checks
+		if (baseType.isArray())
+			baseType = baseType.getComponentType();
+		// If 'base' is single or the '#getReturnType' returns a non 'Object.class' we can use the same operation
+		if (base.isSingle() || !baseType.equals(Object.class)) {
+			if (operation == null) {
+				Object baseObject = base.getSingle(event);
+				if (baseObject == null)
+					return;
+				// If we reached here through 'base.isSingle()', need to ensure the class is not 'Object.class'
+				baseType = baseType.equals(Object.class) ? baseObject.getClass() : baseType;
+				operation = getOperation(baseType, operativeType);
+				if (operation == null) {
+					error(Utils.A(getClassInfoCodeName(baseType)) + " cannot be " + getOperatorString()
+						+ " with " + Utils.a(getClassInfoCodeName(operativeType)) + ".");
+					return;
+				}
+			}
+			Operation<Object, Object, Object> finalOperation = operation;
+			changerFunction = object -> finalOperation.calculate(object, operativeObject);
+		} else {
+			changerFunction = object -> {
+				Operation<Object, Object, Object> thisOperation = getOperation(object.getClass(), operativeType);
+				if (thisOperation != null)
+					return thisOperation.calculate(object, operativeObject);
+				return object;
+			};
```

Comment:
i think a good amount of this can be rewritten to be clear-er and more secure
see ArithmeticChain.get(Event)